storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingSearcher.java

```diff
@@ -0,0 +1,95 @@
+package com.pushtorefresh.storio;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Common class for search and cache indirect type mapping.
+ */
+public class TypeMappingSearcher {
+
+    @Nullable
+    private final Map<Class<?>, ? extends TypeMapping<?>> directTypesMapping;
+
+    @NonNull
+    private final Map<Class<?>, TypeMapping<?>> indirectTypesMappingCache
+            = new ConcurrentHashMap<Class<?>, TypeMapping<?>>();
+
+    public TypeMappingSearcher(@Nullable Map<Class<?>, ? extends TypeMapping<?>> directTypesMapping) {
+        this.directTypesMapping = directTypesMapping;
+    }
+
+    /**
+     * Search algorithm:
+     * Check current class for direct or indirect type mapping from cache.
+     * Walk through all parent classes and interfaces.
+     * If parent type has direct mapping -> we found indirect type mapping!
+     * If current parent type == Object.class or doesn't implement another interfaces ->
+     * there is no indirect type mapping.
+     * Complexity:
+     * O(n) where n is number of parent types of passed type (pretty fast).
+     * <p>
+     * Stop search if root parent is Object.class and doesn't implement another interfaces.
+     *
+     * @return direct or indirect type mapping for passed type, or {@code null}.
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <T> TypeMapping<T> findTypeMappingRecursive(@NonNull final Class<T> type) {
+
+        TypeMapping<T> mapping = getTypeMappingForCurrent(type);
+        if (mapping != null) {
+            return mapping;
+        }
+
+        // Okay, we don't have direct type mapping.
+        // And we don't have cache for indirect type mapping.
+        // Let's find indirect type mapping and cache it!
+
+        Class<?> parentType = type.getSuperclass();
+        if (parentType != null && parentType != Object.class) {
+            mapping = (TypeMapping<T>) findTypeMappingRecursive(parentType);
+            if (mapping != null) {
+                indirectTypesMappingCache.put(type, mapping);
+                return mapping;
+            }
+        }
+
+        // Let's try to find indirect type mapping in own interfaces
+        for (Class<?> ownInterface : type.getInterfaces()) {
+            mapping = (TypeMapping<T>) findTypeMappingRecursive(ownInterface);
+            if (mapping != null) {
+                indirectTypesMappingCache.put(type, mapping);
+                return mapping;
+            }
+        }
+
+        // No type mapping found.
+        return null;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    private <T> TypeMapping<T> getTypeMappingForCurrent(@NonNull final Class<T> type) {
+        if (directTypesMapping == null) {
+            return null;
```

Comment:
Please add comment: `no known type mappings, looks like user forgot to add them` or something… at first sight I though that it's a bug lol